### DISCUSSION TITLE
Has anyone got postal working locally on their computer with docker using Windows WSL?

### DISCUSSION THREAD

**Discussion by ghost** [postalserver/postal#2839] [Help with using Postal]
**What problem am I having?**
I can only seem to get postal working on a real webserver - not locally on my computer with docker.

---

**What have I tried?**
I'm mentally exhausted so I don't have it in me to type out everything I've tried but here's the short versions. The short version is I've tried to get this working using a Dockerfile (builds all the prerequisites) with docker compose, docker compose alone and just a container pulling the ubuntu image and manually following all of the installation steps one-by-one inside the container.

For every combination of things I've tried, I've run into one issue or another - the latest issue is the one below:
```
Postal::ConfigError: No config found at /config/postal.yml (Postal::ConfigError)
```

This happens when trying to run `postal initialize` or `postal start` directly from within the local docker container. I've tried to copy the `/opt/postal/config/postal.yml` file to `/config/postal.yml`, I've tried sim-linking the file from the opt folder as well to the config directory - nothing works. For some reason ruby isn't finding the file in docker though it physically exists when doing `ls -la` for both folders 🙃.

At this point I am just curious if anyone has been successful in making this work locally with docker and how they did it; or if I should just give up on forcing this to work locally.

Note: I am running docker within Windows WSL on Ubuntu (in case that's somehow the reason this has been so hard)

---

**Why do I want this to work locally?**
Well the API for postal is lacking the ability to add HTTP endpoints and routes programmatically. I don't know ruby and learning it just to fit my use-case would take more time than the workaround I am trying to put in place. 

---

**The proposed workaround**
- Use external Maria DB instead of a containerized one (already done, works fine on my actual web server)
- Add HTTP endpoints and routes directly to the external DB from my web application

I'm sure writing to the DB directly will work just fine however, I did not want to do a bunch of testing against the DB actually linked to my email server for this workaround. It would have been nice to be able test everything locally.

As a note, I probably am just lacking the docker knowledge to fully troubleshoot this further than I've been able to - and that's on me.

**Comment by willpower232** [comment]
You should be able to follow the traditional installation process regardless of where you install it.

At the very least, you can reference the installation docker-compose file to get an idea of what goes where https://github.com/postalserver/install/blob/main/templates/docker-compose.yml

**Reply by ghost** [reply]
Thanks for responding Will. I'll give everything another go from scratch later on today.

---

[EDIT] There definitely seems to be some sort of issue relating to docker on windows WSL in later versions of Ubuntu which I believe is contributing to the issue of getting this to work locally. I've been mounting my `docker.sock` file  as a volume to get around the issue noted [here](https://github.com/docker/for-linux/issues/1406). 

I realize now this might be why certain config files can't be found though they exist in the container. It's clear Docker needs to be properly running inside of the container instance - not just mounted from the outside. Will try the troubleshooting steps pointed out by [this user](https://github.com/docker/for-linux/issues/1406#issuecomment-1183487816) before trying to install postal locally again. Just wanted to provide this update in case someone else runs into this. Will also update if I get this working [/EDIT]

---

Well the TDLR here is I don't think docker in docker is really compatible with Ubuntu using WSL for Windows. I fought with this for more hours tonight and I've officially given up on getting this to work locally. Still leaving this here in case someone figures out a solution.
